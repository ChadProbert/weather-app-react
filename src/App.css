/* removed the ability to select text on the app */
.App {
	user-select: none;
}

/* added a transparent background */
.card {
	background-color: rgba(255, 255, 255, 0.103) !important;
}

.headingWeatherApp {
	font-size: 55px;
	margin: 35px;
	padding: 20px;
	border: 5px solid;
	border-radius: 5px;
}

.weather-icon {
	padding-top: 50px;
	width: 100px;
}

.divReadings {
	padding-top: 100px;
}

/* centered loading icon */
.loading {
	position: absolute;
	top: 50%;
	left: 50%;
	margin-right: -50%;
	transform: translate(-50%, -50%);
}

/* search icon */
.fa-search {
	padding-left: 20px;
	padding-right: 20px;
}

/* gave spacing between textfield and the search button */
.btn {
	margin-left: 20px;
}

/* References:

https://www.w3.org/Style/Examples/007/center.en.html

*/
