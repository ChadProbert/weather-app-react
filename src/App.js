import React from 'react'
import "./App.css"
import { useState, useEffect } from "react";
import "bootstrap/dist/css/bootstrap.min.css";
import countries from "i18n-iso-countries";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSpinner } from "@fortawesome/free-solid-svg-icons";
import TextField from "@mui/material/TextField";

countries.registerLocale(require("i18n-iso-countries/langs/en.json"));

export default function App() {
	// stores the response from the API
	const [apiData, setApiData] = useState({});
	// stores the location name that is inside of the input field
	const [getState, setGetState] = useState("Johannesburg");
	// stores a copy of getState
	const [state, setState] = useState("Johannesburg");

	// created a .env.local file which stores the api key as a variable called REACT_APP_API_KEY
	const apiKey = process.env.REACT_APP_API_KEY;
	const apiUrl = `https://api.openweathermap.org/data/2.5/weather?q=${state}&appid=${apiKey}`;

	// fetching data from apiURL and storing inside of apiData state
	useEffect(() => {
		const fetchWeather = async () => {
			const data = await fetch(apiUrl).then((res) => res.json());
			setApiData(data);
		};
		fetchWeather();
	}, [apiUrl]);

	// user input handler functions:

	// stores the input from the user (inside the input field) in getState
	const inputHandler = (event) => {
		setGetState(event.target.value);
	};

	// copies getState to setState
	const submitHandler = () => {
		setState(getState);
	};

	// converting from kelvin (API's default temperature measurement) to degrees celsius
	// conversion formula: kelvin temperature - 273.15 = degrees celsius
	// rounded off to 0 decimal places
	const kelvinToCelsius = (k) => {
		return (k - 273.15).toFixed(0);
	};

	// added Font Awesome spinner icon for when the location search is loading
	const iconSpinner = <FontAwesomeIcon icon={faSpinner} spin />;

	return (
		<div className="App">
			<header className="d-flex justify-content-center align-items-center">
				<h1 className="headingWeatherApp">Weather App</h1>
			</header>

			<div className="container">
				<div className="mt-3 d-flex flex-column justify-content-center align-items-center">
					<span>
						<TextField
							id="standard-basic"
							label="Enter Location"
							variant="standard"
							color="primary"
							onChange={inputHandler}
							value={getState}
						/>
						<button className="btn btn-primary mt-2" onClick={submitHandler}>
							<i className="fas fa-search"></i>
						</button>
					</span>
				</div>
				<div
					className="card mt-3 mx-auto"
					style={{ width: "60vw", height: "60vh" }}
				>
					{apiData.main ? (
						<div className="card-body text-center">
							<img
								src={`http://openweathermap.org/img/w/${apiData.weather[0].icon}.png`}
								alt="weather status icon"
								className="weather-icon"
							/>

							<p className="h2">{kelvinToCelsius(apiData.main.temp)}&deg; C</p>

							<p className="h5">
								<i className="fas fa-map-marker-alt"></i>{" "}
								<strong>{apiData.name}</strong>
							</p>

							<div className="row mt-4 divReadings">
								<div className="col-md-6">
									<p>
										<i className="fas fa-temperature-low "></i>{" "}
										<strong>
											{kelvinToCelsius(apiData.main.temp_min)}&deg; C
										</strong>
									</p>
									<p>
										<i className="fas fa-temperature-high"></i>{" "}
										<strong>
											{kelvinToCelsius(apiData.main.temp_max)}&deg; C
										</strong>
									</p>
								</div>
								<div className="col-md-6">
									<p>
										{" "}
										<strong>{apiData.weather[0].main}</strong>
									</p>
									<p>
										<strong>
											{" "}
											{countries.getName(apiData.sys.country, "en", {
												select: "official",
											})}
										</strong>
									</p>
								</div>
							</div>
						</div>
					) : (
						<h1 className="loading">{iconSpinner}</h1>
					)}
				</div>
			</div>
		</div>
	);
}

/* References: 

https://openweathermap.org/

https://dev.to/imshines/a-simple-weather-app-using-react-and-openweathermap-api-10m2

https://www.vedantu.com/chemistry/kelvin-to-celsius

https://www.npmjs.com/package/i18n-iso-countries

https://stackoverflow.com/questions/49108136/importing-env-variable-react-front-end

https://react-bootstrap.github.io/getting-started/introduction/

https://fontawesome.com/v5.15/how-to-use/on-the-web/using-with/react

https://mui.com/components/text-fields/

*/
