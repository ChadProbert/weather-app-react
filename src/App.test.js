import React from "react";
import renderer from "react-test-renderer"
import { render } from "@testing-library/react";
import App from "./App";

describe("testing snapshot", () => {

  test("snapshot for App component", () => {
    /* converts the code inside of the App component to JSON format 
    and is stored as a snapshot inside the project directory */
    const renderedComponent = renderer.create(<App />).toJSON()
    expect(renderedComponent).toMatchSnapshot()
  })

})

global.fetch = jest.fn(() =>
	Promise.resolve({
		json: () =>
			Promise.resolve({
				name: "Johannesburg",
				country: "ZA",
			}),
	})
);

describe("App", () => {
	it("loads Johannesburg weather", () => {
		const { container } = render(<App />);
		expect(container.innerHTML).toMatch("Johannesburg");
	});
});

/* References: 

https://www.youtube.com/watch?v=qOaGoujjc3M

https://testing-library.com/docs/react-testing-library/intro/

https://jestjs.io/docs/asynchronous

https://www.youtube.com/watch?time_continue=64&v=yTZ-txdrHdY&feature=emb_title

https://api.openweathermap.org/data/2.5/weather?q=Johannesburg&appid=ed11dcfaa8049f2c5c3c5985abd88e40

*/
