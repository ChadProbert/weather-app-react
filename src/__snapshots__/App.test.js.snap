// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing snapshot snapshot for App component 1`] = `
<div
  className="App"
>
  <header
    className="d-flex justify-content-center align-items-center"
  >
    <h1
      className="headingWeatherApp"
    >
      Weather App
    </h1>
  </header>
  <div
    className="container"
  >
    <div
      className="mt-3 d-flex flex-column justify-content-center align-items-center"
    >
      <span>
        <div
          className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={true}
            htmlFor="standard-basic"
            id="standard-basic-label"
          >
            Enter Location
          </label>
          <div
            className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              disabled={false}
              id="standard-basic"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="Johannesburg"
            />
          </div>
        </div>
        <button
          className="btn btn-primary mt-2"
          onClick={[Function]}
        >
          <i
            className="fas fa-search"
          />
        </button>
      </span>
    </div>
    <div
      className="card mt-3 mx-auto"
      style={
        Object {
          "height": "60vh",
          "width": "60vw",
        }
      }
    >
      <h1
        className="loading"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-spinner fa-w-16 fa-spin "
          data-icon="spinner"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </h1>
    </div>
  </div>
</div>
`;
